apps/rush-lib/src/logic/buildCache/AzureStorageBuildCacheProvider.ts

```diff
@@ -0,0 +1,378 @@
+  BlobServiceClient,
+  BlockBlobClient,
+  ContainerClient,
+  generateBlobSASQueryParameters,
+  SASQueryParameters,
+  ServiceGetUserDelegationKeyResponse,
+  UserDelegationKey
+} from '@azure/storage-blob';
+import { DeviceCodeCredential, DeviceCodeInfo } from '@azure/identity';
+
+import { EnvironmentConfiguration, EnvironmentVariableNames } from '../../api/EnvironmentConfiguration';
+import { RushGlobalFolder } from '../../api/RushGlobalFolder';
+import {
+  BuildCacheProviderCredentialCache,
+  IBuildCacheProviderCredentialCacheEntry
+} from './BuildCacheProviderCredentialCache';
+import { URLSearchParams } from 'url';
+import { RushConstants } from '../RushConstants';
+import { Utilities } from '../../utilities/Utilities';
+
+export type AzureEnvironmentNames =
+  | 'AzureCloud'
+  | 'AzureChinaCloud'
+  | 'AzureUSGovernment'
+  | 'AzureGermanCloud';
+export enum AzureEnvironment {
+  AzureCloud,
+  AzureChinaCloud,
+  AzureUSGovernment,
+  AzureGermanCloud
+}
+
+export interface IAzureStorageBuildCacheProviderOptions extends IBuildCacheProviderBaseOptions {
+  storageContainerName: string;
+  storageAccountName: string;
+  azureEnvironment?: AzureEnvironment;
+  blobPrefix?: string;
+  isCacheWriteAllowed: boolean;
+  rushGlobalFolder: RushGlobalFolder;
+}
+
+const SAS_TTL: number = 7 * 24 * 60 * 60 * 1000; // Seven days
+
+export class AzureStorageBuildCacheProvider extends BuildCacheProviderBase {
+  private readonly _storageAccountName: string;
+  private readonly _storageContainerName: string;
+  private readonly _azureEnvironment: AzureEnvironment;
+  private readonly _blobPrefix: string | undefined;
+  private readonly _isCacheWriteAllowed: boolean;
+  private readonly _rushGlobalFolder: RushGlobalFolder;
+  private __credentialCacheId: string | undefined;
+
+  private _containerClient: ContainerClient | undefined;
+
+  public constructor(options: IAzureStorageBuildCacheProviderOptions) {
+    super(options);
+    this._storageAccountName = options.storageAccountName;
+    this._storageContainerName = options.storageContainerName;
+    this._azureEnvironment = options.azureEnvironment || AzureEnvironment.AzureCloud;
+    this._blobPrefix = options.blobPrefix;
+    this._isCacheWriteAllowed = options.isCacheWriteAllowed;
+    this._rushGlobalFolder = options.rushGlobalFolder;
+  }
+
+  private get _credentialCacheId(): string {
+    if (!this.__credentialCacheId) {
+      let serializedAzureEnvironmentName: string;
+      switch (this._azureEnvironment) {
+        case AzureEnvironment.AzureCloud: {
+          serializedAzureEnvironmentName = 'AzureCloud';
+          break;
+        }
+
+        case AzureEnvironment.AzureChinaCloud: {
+          serializedAzureEnvironmentName = 'AzureChinaCloud';
+          break;
+        }
+
+        case AzureEnvironment.AzureUSGovernment: {
+          serializedAzureEnvironmentName = 'AzureUSGovernment';
+          break;
+        }
+
+        case AzureEnvironment.AzureGermanCloud: {
```

Comment:
Why is this mapping not provided by the AWS SDK?  It seems weird for Rush to have to maintain this table.

Reply:
Yeah it is kinda. Updated this.

Reply:
Assuming you were referring to the Azure Storage SDK.